lib/server/server.mjs

```diff
@@ -0,0 +1,406 @@
+      stream,
+      pipe: stream.pipe(ndjson.parse()),
+      handler: data => this.#handleMessage(data, streamContext)
+    })
+
+    this.#logger.debug({ id: streamContext.id, stream }, 'Connecting stream')
+
+    streamContext.pipe.on('data', streamContext.handler)
+    streamContext.stream.on('error', err => this.emit('error',
+      new StreamingError(stream, err))
+    )
+    streamContext.pipe.on('error', err => this.emit('error',
+      new StreamingError(stream, err))
+    )
+
+    stream.on('close', () => {
+      this.#logger.debug({ id: streamContext.id },
+        'Stream closed, disconnecting')
+      this.disconnect(stream)
+    })
+
+    this.#streams.set(stream, streamContext)
+    this.emit('connect', stream)
+  }
+
+  /**
+  * @summary Disconnect the stream, no longer listening to messages coming from
+  * it.
+  *
+  * @description Additionally, a 'disconnect' event is also emitted with the
+  * disconnected stream as event data.
+  *
+  * @param {stream.Duplex} stream Stream
+  */
+  disconnect (stream) {
+    const streamContext = this.#streams.get(stream)
+
+    if (!streamContext) { return }
+
+    this.#logger.debug({ id: streamContext.id, stream }, 'Disconnecting stream')
+    streamContext.pipe.off('data', streamContext.handler)
+    this.#streams.delete(stream)
+    this.emit('disconnect', stream)
+  }
+
+  /**
+  * @param {Message} message
+  * @param {StreamContext} streamContext
+  */
+  #handleMessage (message, streamContext) {
+    const logger = this.#logger.child({ stream: streamContext.id })
+    logger.info({ message }, 'Received message')
+
+    // Validate message
+    try {
+      assert(message?.header, 'Missing header!')
+      assert(message?.header?.correspondenceId?.length > 0,
+        'Missing correspondence id!')
+      assert(message?.header?.subject?.length > 0, 'Missing subject!')
+    } catch (err) {
+      logger.error({ message }, 'Received invalid message, emitting error')
+
+      // Emit error event and bail
+      this.emit('error',
+        new InvalidMessageError(streamContext.stream, message, err.message)
+      )
+
+      return
+    }
+
+    // Prepare environment
+    const request = new Request({
+      ...message,
+      stream: streamContext.stream
+    })
+
+    const response = new Response({
+      header: message.header,
+      stream: streamContext.stream
+    })
+
+    const context = {}
+
+    // Call handlers
+    const handlers =
+      this.#handlers.get(message.header.subject) ?? this.#defaultHandlers
+
+    handlers === this.#defaultHandlers &&
+      logger.warn({ subject: message.header.subject },
+        'Message subject unknown, calling default handlers')
+
+    for (const handler of handlers) {
+      try {
+        // TODO: Support async handlers
+        handler(request, response, context)
+      } catch (err) {
+        logger.warn({ err }, 'Caught exception processing message')
+        for (let i = 0; i < this.#exceptionHandlers.length; ++i) {
```

Comment:
Somehow I'm getting braincramps from seeing one-letter variable names in "modern code". I know, `i` as a loop index is kinda de-facto, but still, maybe `let index = 0`? You can essentially ignore this comment, it's most possibly just my own weirdness. :D